File: Python/111. Minimum Depth of Binary Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        if not root.left and not root.right:
            return 1
        elif not root.left:
            return 1 + self.minDepth(root.right)
        elif not root.right:
            return 1 + self.minDepth(root.left)
        else:
            return 1 + min(self.minDepth(root.left), self.minDepth(root.right))

# class Solution:
#     def minDepth(self, root: Optional[TreeNode]) -> int:
#         res = 0
#         que = []
#         if root:
#             que.append(root)
#         while que:
#             res += 1
#             size = len(que)
#             for i in range(size):
#                 cur = que.pop(0)
#                 if cur.left:
#                     que.append(cur.left)
#                 if cur.right:
#                     que.append(cur.right)
#                 if not cur.left and not cur.right:
#                     return res
#         return res

File: C++/111. Minimum Depth of Binary Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == NULL) return 0;
        if (!root->left && !root->right) return 1;
        else if (root->left && !root->right) return 1 + minDepth(root->left);
        else if (root->right && !root->left) return 1 + minDepth(root->right);
        else {
            int left = minDepth(root->left);
            int right = minDepth(root->right);
            return min(left, right) + 1;
        }
    }
};

// class Solution {
// public:
//     int minDepth(TreeNode* root) {
//         queue<TreeNode*> que;
//         if (root) que.push(root);
//         int res = 0;
//         while (!que.empty()){
//             res++;
//             int size = que.size();
//             for (int i=0; i<size; i++){
//                 TreeNode* cur = que.front();
//                 que.pop();
//                 if (!cur->left && !cur->right) return res;
//                 if (cur->left) que.push(cur->left);
//                 if (cur->right) que.push(cur->right);
//             }
//         }
//         return res;
//     }
// };